inferelator_ng/bbsr_python.py

```diff
@@ -0,0 +1,301 @@
+import pandas as pd
+import numpy as np
+import itertools
+from itertools import compress
+import math
+from scipy import special
+import multiprocessing
+from functools import partial
+import os, sys
+
+# Wrapper function for BBSRforOneGene that's called in BBSR
+gx, gy, gpp, gwm, gns = None, None, None, None, None
+def BBSRforOneGeneWrapper(ind): return BBSRforOneGene(ind, gx, gy, gpp, gwm, gns)
+
+def BBSR(X, Y, clr_mat, nS, no_pr_val, weights_mat, prior_mat, kvs, rank, ownCheck):
+    G = Y.shape[0] # number of genes
+    genes = Y.index.values.tolist()
+    K = X.shape[0]  # max number of possible predictors (number of TFs)
+    tfs = X.index.values.tolist()
+
+    # Scale and permute design and response matrix
+    X = ((X.transpose() - X.transpose().mean()) / X.transpose().std(ddof=1)).transpose()
+    Y = ((Y.transpose() - Y.transpose().mean()) / Y.transpose().std(ddof=1)).transpose()
+
+    weights_mat = weights_mat.loc[genes,tfs]
+    clr_mat = clr_mat.loc[genes, tfs]
+    prior_mat = prior_mat.loc[genes, tfs]
+    # keep all predictors that we have priors for
+    pp = pd.DataFrame(((prior_mat.ix[:,:] != 0)|(weights_mat.ix[:,:]!=no_pr_val)) & ~pd.isnull(clr_mat))
+    mask = clr_mat == 0
+    # for each gene, add the top nS predictors of the list to possible predictors
+    clr_mat[mask] = np.nan
+
+    for ind in range(0,G):
+
+        clr_na = len(np.argwhere(np.isnan(clr_mat.ix[ind,])).flatten().tolist())
+        clr_w_na = np.argsort(clr_mat.ix[ind,].tolist())
+
+        if clr_na>0:
+            clr_order = clr_w_na[:-clr_na][::-1]
+        else:
+            clr_order = clr_w_na[:][::-1]
+        pp.ix[ind, clr_order[0:min(K, nS, len(clr_order))]] = True
+
+    preds = np.intersect1d(genes, tfs)
+    subset = pp.ix[preds,preds].values
+    np.fill_diagonal(subset,False)
+    pp=pp.set_value(preds, preds, subset)
+
+    out_list=[]
+
+    global gx, gy, gpp, gwm, gns
+    gx, gy, gpp, gwm, gns = X, Y, pp, weights_mat, nS
+    # Here we illustrate splitting a simple loop, but the same approach
+    # would work with any iterative control structure, as long as it is
+    # deterministic.
+    s = []
+    limit = G
+    for j in range(limit):
+        if not ownCheck or ownCheck.next():
+            # busy work
+            s.append(BBSRforOneGeneWrapper(j))
```

Comment:
This removes the need for globals:
`s.append(BBSRforOneGene(X, Y, pp, weights_mat, nS))`